### PR TITLE
Add JDBC device registry

### PR DESCRIPTION
**Pull request by ctron**
The CI is expected to fail on this one as well, as it misses changes from: https://github.com/eclipse/hono/pull/2157 and https://github.com/eclipse/hono/pull/2158

### CODE REVIEW

**Comment by ctron** [comment]
Rebased

File: tests/pom.xml [hunk 1]
```diff
@@ -267,6 +286,74 @@ Test cases are run against Docker images of Hono server + (Apache Qpid Dispatch
...
+                          <includes>
+                            <include>**/*</include>
+                          </includes>
+                        </fileSet>
+                      </inline>
+                    </assembly>
+                    <runCmds>
+
+                      <run>microdnf -y install java-11-openjdk-headless</run>
+                      <run>curl -sL https://repo1.maven.org/maven2/com/h2database/h2/1.4.200/h2-1.4.200.jar -o h2.jar</run>
```
Comment (sophokles73) [inline tests/pom.xml:314]: can we use the version defined in `bom/POM.xml` here instead?

File: tests/pom.xml [hunk 2]
```diff
@@ -314,6 +401,7 @@ Test cases are run against Docker images of Hono server + (Apache Qpid Dispatch
                     </network>
                     <memorySwap>524288000</memorySwap>
                     <memory>524288000</memory>
+                    <memory>524288000</memory>
```
Comment (sophokles73) [inline tests/pom.xml:404]: this seems to be a duplicate

**Review by sophokles73 — COMMENTED** [review]
looks very good :-)

**Comment by ctron** [comment]
Ok, I rebased, amended the feedback, and (most likely) fixes the test. Which was due to the fact that the "max bcrypt iterations" variable was renamed to "cost factor". But let's wait for travis to check.

**Comment by sophokles73** [comment]
@ctron is there anything else I can help with? FMPOV the only thing that prevents this PR from merging is the failing unit test ...

**Comment by fkaltner** [comment]
> @ctron is there anything else I can help with? FMPOV the only thing that prevents this PR from being merged is the failing unit test ...

@ctron I am also happy to help, if there's anything I could do FYPOV. @dejanb would like to wait with merging the pull request for the status property (https://github.com/eclipse/hono/pull/2189) until this PR is merged...

File: services/device-registry-jdbc/src/main/java/org/eclipse/hono/deviceregistry/jdbc/impl/CredentialsServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+        }
+
+        if (!secret.getBoolean(CredentialsConstants.FIELD_ENABLED, true)) {
+            return false;
+        }
+
+        if (!validTime(secret, CredentialsConstants.FIELD_SECRETS_NOT_BEFORE, Instant::isAfter)) {
+            return false;
+        }
+        if (!validTime(secret, CredentialsConstants.FIELD_SECRETS_NOT_AFTER, Instant::isBefore)) {
```
Comment (kaniyan) [inline services/device-registry-jdbc/src/main/java/org/eclipse/hono/deviceregistry/jdbc/impl/CredentialsServiceImpl.java:106]: @sophokles73 @ctron 

Sorry for raising this comment now. I noticed this just now with the merge of the PR #2241 for adjusting the tests.

The Credentials API docs says:

> Adapters MUST only consider secrets for authentication which
> 
>     have their not-before property set to either null or the current or a past point in time and
>     have their not-after property set to either null or the current or a future point in time.

AFAIK the credentials API includes all the secrets irrespective of the _not-before_ & _not-after_ property. Based on the above quotes, the adapters are responsible for filtering them out. Various _AuthProvider_ implementations already do this by using `CredentialsObject.getCandidateSecrets(...)` where the secrets are already being filtered out based on the validity period. IMHO we need not filter secrets based on the validity period in the credentials API implementation. I don't think we need to adapt the tests as in #2241

**Review by kaniyan — CHANGES_REQUESTED** [review]

**Pull request merged**